source/connection-monitoring-and-pooling/connection-monitoring-and-pooling.rst

```diff
@@ -355,29 +397,85 @@ Before a connection can be returned outside of the pool, it must be “set up”
 Closing a Connection (Internal Implementation)
 ----------------------------------------------
 
-When a Connection is closed, it MUST first be marked as closed, removing it from being counted as available, in use, or setup. Once the connection is marked as closed, the Connection can perform whatever teardown is necessary. The Driver SHOULD perform the teardown in a non-blocking manner.
+When a `Connection <#connection>`_ is closed, it MUST first be marked as "closed",
+removing it from being counted as "available" or "in use". One that is
+complete, the `Connection <#connection>`_ can perform whatever teardown is
+necessary to close its underlying socket. The Driver MUST perform this
+teardown in a non-blocking manner, such as via the use of a background
+thread or async I/O.
 
 .. code::
 
-    mark connection as closed
+    decrement total connection count
+    if connection state is "available":
+      decrement available connection count
+    set connection state to "closed"
     emit ConnectionClosedEvent
 
-    # The following can happen at a later time
-    connection.close()
+    # The following can happen at a later time (i.e. in background
+    # thread) or via non-blocking I/O.
+    connection.socket.close()
+
+Marking a Connection as Available (Internal Implementation)
+-----------------------------------------------------------
+
+A `Connection <#connection>`_ is "available" if it is able to be checked out. A
+`Connection <#connection>`_ MUST NOT be marked as "available" until it has been
+established. The pool MUST keep track of the number of currently
+available `Connections <#connection>`_.
+
+.. code::
+
+   increment available connection count
+   set connection state to "available"
+   add connection to availableConnections
+
+
+Populating the Pool with a Connection (Internal Implementation)
+---------------------------------------------------------------
+
+If minPoolSize is set, the pool MUST ensure that it always has at
+least minPoolSize total `Connections <#connection>`_. In order to achieve this, it
+needs to be able to create `Connections <#connection>`_ that begin as "available"
+instead of "in use" (i.e. "populate the pool"). Performing this
```

Comment:
The "beginning as available vs in use" distinction is confusing to me. Conceptually all connections go through both states, i.e. each connection follows pending -> available -> in use. A driver could create all connections in background and block main thread until there is a connection created in the background, a driver could create connections in foreground only and a driver could do both. In Ruby we considered all of these approaches and settled on the hybrid one which provides the best responsiveness to application needs. But this is an implementation detail and I am of the opinion it does not belong in the spec.

Looking at this section (populating the pool), perhaps it can be removed completely because  I believe all of its content is already covered elsewhere in the spec.